### PR TITLE
new success api and improved file handling

### PR DESCRIPTION
**Pull request by xithan**

### CODE REVIEW

File: card/lib/card.rb [hunk 1]
```diff
@@ -53,8 +53,9 @@ def self.serializable_attr_accessor *args
   attr_accessor :follower_stash
 
 
-  define_callbacks :approve, :store, :stored, :extend, :subsequent
+  define_callbacks :prepare, :approve, :store, :stored, :extend, :subsequent, :select_action
 
+  before_validation :prepare
```
Comment (ethn) [inline card/lib/card.rb]: I had sort of been leaning this way a bit.  Kind of feels like we need to put all the events on post-its and arrange and re-arrange until everything makes sense...

Comment (GerryG) [inline card/lib/card.rb]: Then put the result on wagn.org so devs can find it ;)

File: card/mod/01_history/set/all/actions.rb [hunk 2]
```diff
@@ -1,13 +1,17 @@
 # -*- encoding : utf-8 -*-
 
 def select_action_by_params params
-  action = find_action_by_params(params) and self.selected_action_id = action.id
+  if (action = find_action_by_params(params))
+    run_callbacks :select_action do
```
Comment (ethn) [inline card/mod/01_history/set/all/actions.rb]: this is our first "READ" event.

**Comment by ethn** [comment]
Is it safe to assume that decks will break unless their js machine output is updated?  If so, we probably need to either re-use our short term hack (deleting the machine output in a migration) or implement the long-term solution we discussed (making output files smart about file changes).

I'm cool with the short-term one for now.

**Comment by xithan** [comment]
That's right. I'll add the short-term solution

File: card/db/migrate_core_cards/20150824135418_update_file_history.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  def up
+    Card.search(:type=>[:in, 'file', 'image']).each do |card|
+      card.actions.each do |action|
+        if (content_change = action.change_for(:db_content).first)
+          original_filename, file_type, action_id, mod  = content_change.value.split("\n")
+          if file_type.present? && action_id.present?
+            if mod.present?
+              content_change.update_attributes! :value=>":#{card.codename}/#{mod}#{::File.extname(original_filename)}"
+            else
+              content_change.update_attributes! :value=>"~#{card.id}/#{action_id}#{::File.extname(original_filename)}"
```
Comment (ethn) [inline card/db/migrate_core_cards/20150824135418_update_file_history.rb]: nitpick.  could do:
````
value = mod.present? ? A : B
content_change.update_attributes! :value=>value
````

File: wagn/rails/controllers/card_controller.rb [hunk 4]
```diff
@@ -184,50 +194,23 @@ def card_redirect url
...
-        self.params = new_params
+      if @success.soft_redirect?
+        self.params = @success.params
       else
-        self.params.merge! new_params # #need tests. insure we get slot, main...
+        self.params.merge! @success.params # #need tests. insure we get slot, main...
       end
+      @card = @success.target
+      @card.select_action_by_params params
       show
```
Comment (ethn) [inline wagn/rails/controllers/card_controller.rb]: love this!

File: wagn/rails/controllers/card_controller.rb [hunk 5]
```diff
@@ -44,7 +48,9 @@ def update
 
   def delete
     discard_locations_for card #should be an event
-    params[:success] ||= 'REDIRECT: *previous'
+    if @success.target == card
+      @success.target = :previous
+    end
```
Comment (ethn) [inline wagn/rails/controllers/card_controller.rb]: I suppose we could put this in an event now...  (though it's already cleaner now than it was)

File: card/lib/card/success.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      else
+        super
+      end
+    end
+
+    def session
+      Card::Env.session
+    end
+  end
+
```
Comment (ethn) [inline card/lib/card/success.rb]: drastic improvement

**Pull request merged**